src/agent/metrics.rs

```diff
@@ -0,0 +1,376 @@
+use {
+    super::{
+        solana::oracle::PriceAccount,
+        store::{
+            global::{
+                AllAccountsData,
+                AllAccountsMetadata,
+                Lookup,
+                PriceAccountMetadata,
+            },
+            local::{
+                Message,
+                PriceInfo,
+            },
+        },
+    },
+    chrono::NaiveDateTime,
+    pyth_sdk::{
+        Identifier,
+        PriceIdentifier,
+    },
+    slog::Logger,
+    solana_sdk::pubkey::Pubkey,
+    std::{
+        collections::{
+            BTreeMap,
+            BTreeSet,
+            HashMap,
+            HashSet,
+        },
+        net::SocketAddr,
+        sync::Arc,
+        time::Instant,
+    },
+    tokio::sync::{
+        mpsc,
+        oneshot,
+        Mutex,
+    },
+    typed_html::{
+        dom::DOMTree,
+        html,
+        text,
+    },
+    warp::{
+        hyper::StatusCode,
+        reply::{self,},
+        Filter,
+        Rejection,
+        Reply,
+    },
+};
+
+/// Internal metrics server state, holds state needed for serving
+/// dashboard and metrics.
+pub struct MetricsServer {
+    /// Used to pull the state of all symbols in local store
+    local_store_tx:         mpsc::Sender<Message>,
+    global_store_lookup_tx: mpsc::Sender<Lookup>,
```

Comment:
i presume in the future you're going to create another component that aggregates metrics over time (like gets sent info on each update or whatever to track in accumulators), and the data from that component will be accessible here as well?

I think the current information here is a good start to a dashboard, but I'm pretty sure that people will want to see metrics over some time interval. Most problems (like dropped RPC requests, etc) will happen sporadically and you'll want historical data to see whether or not they've been happening.

Reply:
The main piece of Rust boilerplate for adding metrics is that the Rust Prometheus crate does not let you discover existing metrics easily from global metrics state. This state is kept in a `Registry` object, which is the top-level container used to gather and export formatted metrics in bulk. The preferred method for creating/referencing metrics is linking counter/gauge objects to the registry, and using the objects themselves as references to individual metrics, presumably right beside other state variables. 

This pattern was reasonably easy for the attester, where we focused on the overall performance of the system and got away with storing a handful of static metrics as globals.

In the agent, we'll likely need to add the counter objects per-symbol dynamically. I plan to extend global/local store state to hold the new objects and update them with incoming Update* messages. I'ii take good a look at Richard's comments and screenshots to get an idea of what could work well.